### ISSUE TITLE
Remove ability to add a template part in the post editor

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#44397]
### Description

Remove ability to add a template part in the post editor, including copy and pasting. This was previous in place https://github.com/WordPress/gutenberg/pull/37157 but seems to now be allowed to add once more.  I can replicate this with 6.1 beta 1. 

### Step-by-step reproduction instructions

1. Use GB 14.1.1 
2. Open post editor
3. Start to add a template part block however you'd like.
4. Notice that you can.
5. Deactivate Gutenberg
6. Open post editor
7. Try to add a template part block and notice you can't

Separately, notice that you can copy/paste a template part block from the template editor into a post editor, with or without Gutenberg installed, causing multi entity saving to appear in the post editor. 

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/26996883/191822627-f0276537-8fb7-4b26-9b65-b0807d471619.mov

### Environment info

- WordPress 6.1 beta 1
- Gutenberg 14.1.1 both activate and deactivated
- TT2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by ramonjd** [comment]
I only got so far as to realize that `rootClientId` returns `null` at this point in the template part block init(): https://github.com/WordPress/gutenberg/blob/4f7c19e37d55bb6a419cc8eca94c634f1e71852e/packages/block-library/src/template-part/index.js#L67

So the tests are failing and the filter will return `true` (allow the block to be inserted)

I can't see much that might have changed in the meantime since https://github.com/WordPress/gutenberg/pull/37157 🤔  cc @adamziel who might know straight away

**Comment by Mamaduka** [comment]
I think the filter stopped working after #42258 got merged.

## Testing
* ✅  The commit before merging the filter is working - 9620bae6fef4fde7cc2b7833f416e240207cda29
* ❌  The merge commit the filter stops working - 4f7c19e37d55bb6a419cc8eca94c634f1e71852e

It might be a good idea to audit other blocks that are using [side-effects](https://github.com/WordPress/gutenberg/blob/trunk/packages/side-effects.md) - `navigation-link`, `query` and `post-terms`.

@gziolo, any ideas why the filters might stop working?

**Comment by gziolo** [comment]
@Mamaduka, I guess it means that the filter should get applied globally rather than only when the Template Part gets registered. It might be related to the order of execution, too.

**Comment by adamziel** [comment]
> I can't see much that might have changed in the meantime since #37157 🤔 cc @adamziel who might know straight away

Unfortunately I don't :( @Mamaduka and @gziolo seem to be on the right track, though. It sounds like `canInsertBlockTypeUnmemoized` receives null as `rootClientId` on some occasions. It would be nice to have some assertions in place, e.g. `assert( rootClientId )` – that way we'd find such problems right away.

**Comment by Mamaduka** [comment]
> I guess it means that the filter should get applied globally rather than only when the Template Part gets registered. It might be related to the order of execution, too.

That might be the case for filter side effects. Would it be okay to add filters directly in the `init.js` file but outside the `init` method?

**Comment by gziolo** [comment]
> > I guess it means that the filter should get applied globally rather than only when the Template Part gets registered. It might be related to the order of execution, too.
> 
> That might be the case for filter side effects. Would it be okay to add filters directly in the `init.js` file but outside the `init` method?

`init.js` exists only for 3-rd party usage when devs want to init individual blocks. In the case of WordPress core, we call `init` from `index.js. 

I think I identified what has changed:

The filter from the block library got applied first, and then it would get overridden by the following version in post editor:

https://github.com/WordPress/gutenberg/blob/7829913ae117dfb561d14c600eea7b281afd6556/packages/edit-post/src/index.js#L83-L97

Now, the filter registered with the block goes after. The fix should be as simple as moving the code above after `registerCoreBlocks` call in the file.

**Comment by Mamaduka** [comment]
> The filter from the block library got applied first, and then it would get overridden by the following version in post editor

@gziolo, I'm not sure if that's the case.

Even if I move the edit post filter after the `registerCoreBlocks` call or comment it out, I can still insert the Template Parts block.

**Comment by ramonjd** [comment]
> The fix should be as simple as moving the code above after registerCoreBlocks call in the file.

Hmm, this appears to work for me, thanks everyone!

I've thrown up a PR just in case it works for everyone else too 🤣  https://github.com/WordPress/gutenberg/pull/44480
